ddl/ddl_api.go

```diff
@@ -4891,6 +4891,100 @@ func (d *ddl) DropIndex(ctx sessionctx.Context, ti ast.Ident, indexName model.CI
 	return errors.Trace(err)
 }
 
+func (d *ddl) DropIndexes(ctx sessionctx.Context, ti ast.Ident, specs []*ast.AlterTableSpec) error {
+	is := d.infoHandle.Get()
+	schema, ok := is.SchemaByName(ti.Schema)
+	if !ok {
+		return errors.Trace(infoschema.ErrDatabaseNotExists)
+	}
+	t, err := is.TableByName(ti.Schema, ti.Name)
+	if err != nil {
+		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ti.Schema, ti.Name))
+	}
+
+	existsSpec := make(map[string]bool, len(specs))
+	indexNames := make([]model.CIStr, 0, len(specs))
+	for _, spec := range specs {
+		var indexName model.CIStr
+		switch spec.Tp {
+		case ast.AlterTableDropIndex:
+			indexName = model.NewCIStr(spec.Name)
+		case ast.AlterTableDropPrimaryKey:
+			indexName = model.NewCIStr(mysql.PrimaryKeyName)
+		default:
+			return nil
+		}
+
+		indexInfo := t.Meta().FindIndexByName(indexName.L)
+		var isPK bool
+		if indexName.L == strings.ToLower(mysql.PrimaryKeyName) &&
+			// Before we fixed #14243, there might be a general index named `primary` but not a primary key.
+			(indexInfo == nil || indexInfo.Primary) {
+			isPK = true
+		}
+		if isPK {
+			if !config.GetGlobalConfig().AlterPrimaryKey {
+				return ErrUnsupportedModifyPrimaryKey.GenWithStack("Unsupported drop primary key when alter-primary-key is false")
+			}
+			// If the table's PKIsHandle is true, we can't find the index from the table. So we check the value of PKIsHandle.
+			if indexInfo == nil && !t.Meta().PKIsHandle {
+				return ErrCantDropFieldOrKey.GenWithStack("Can't DROP 'PRIMARY'; check that column/key exists")
+			}
+			if t.Meta().PKIsHandle {
+				return ErrUnsupportedModifyPrimaryKey.GenWithStack("Unsupported drop primary key when the table's pkIsHandle is true")
+			}
+			if t.Meta().IsCommonHandle {
+				return ErrUnsupportedModifyPrimaryKey.GenWithStack("Unsupported drop primary key when the table is using clustered index")
+			}
+		}
+		if indexInfo == nil {
+			err = ErrCantDropFieldOrKey.GenWithStack("index %s doesn't exist", indexName)
+			if spec.IfExists {
+				ctx.GetSessionVars().StmtCtx.AppendNote(err)
+				continue
+			}
+			return err
+		}
+
+		// Check for drop index on auto_increment column.
+		err = checkDropIndexOnAutoIncrementColumn(t.Meta(), indexInfo)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if _, ok := existsSpec[indexName.L]; ok {
+			return ErrCantDropFieldOrKey.GenWithStack("Can't DROP '%s'; check that column/key exists", indexName)
```

Comment:
```suggestion
			return ErrCantDropFieldOrKey.GenWithStack("index %s doesn't exist", indexName)
```
`Can't DROP '%s'; check that column/key exists` is reported only when `indexInfo == nil && !t.Meta().PKIsHandle`.

Reply:
What if `ifExists==true`? It should just continue, not return.

Reply:
The DropIndex and DropIndexes have been refactored.